apns2/client.py

```diff
@@ -1,39 +1,156 @@
         server = 'api.development.push.apple.com' if use_sandbox else 'api.push.apple.com'
         port = 2197 if use_alternative_port else 443
         ssl_context = init_context()
         ssl_context.load_cert_chain(cert_file)
-        self.__connection = HTTP20Connection(server, port, ssl_context=ssl_context)
+        self._connection = HTTP20Connection(server, port, ssl_context=ssl_context)
+        self._max_concurrent_streams = None
+        self._previous_server_max_concurrent_streams = None
+
+    def send_notification(
+        self,
+        token_hex,
+        notification,
+        topic,
+        priority=NotificationPriority.Immediate
+    ):
+        stream_id = self.send_notification_async(token_hex, notification, topic, priority)
+        result = self.get_notification_result(stream_id)
+        if result != 'Success':
+            raise exception_class_for_reason(result)
 
-    def send_notification(self, token_hex, notification, priority=NotificationPriority.Immediate, topic=None):
-        json_payload = dumps(notification.dict(), ensure_ascii=False, separators=(',', ':')).encode('utf-8')
+    def send_notification_async(
+        self,
+        token_hex,
+        notification,
+        topic,
+        priority=NotificationPriority.Immediate
+    ):
+        json_payload = json.dumps(
+            notification.dict(),
+            ensure_ascii=False,
+            separators=(',', ':')
+        ).encode('utf-8')
 
-        headers = {
-            'apns-priority': priority.value
-        }
-        if topic:
-            headers['apns-topic'] = topic
+        headers = {'apns-topic': topic}
+        if priority != DEFAULT_APNS_PRIORITY:
+            headers['apns-priority'] = priority.value
 
         url = '/3/device/{}'.format(token_hex)
-        stream_id = self.__connection.request('POST', url, json_payload, headers)
-        resp = self.__connection.get_response(stream_id)
-        if resp.status != 200:
-            raw_data = resp.read().decode('utf-8')
+        stream_id = self._connection.request('POST', url, json_payload, headers)
+        return stream_id
+
+    def get_notification_result(self, stream_id):
+        response = self._connection.get_response(stream_id)
+        if response.status == 200:
+            return 'Success'
+        else:
+            raw_data = response.read().decode('utf-8')
             data = json.loads(raw_data)
-            raise exception_class_for_reason(data['reason'])
+            return data['reason']
+
+    def send_notification_batch(
+        self,
+        tokens,
+        notification,
+        topic,
+        priority=NotificationPriority.Immediate
+    ):
+        '''
+        Send a notification to a list of tokens in batch. Instead of sending a synchronous request
+        for each token, send multiple requests concurrently. This is done on the same connection,
+        using HTTP/2 streams (one request per stream).
```

Comment:
I'm not sure this is the best interface, specifically in terms of the return value, which only returns counts instead of returning a dict of token: outcome.  The caller might want to know which tokens were unregistered, for example.

You could also change this method so it returns a dict, and define a wrapper method that calls this method, logs the results, and returns the outcome counts.